src/node.cc

```diff
@@ -1303,86 +1303,187 @@ void AddPromiseHook(v8::Isolate* isolate, promise_hook_func fn, void* arg) {
   env->AddPromiseHook(fn, arg);
 }
 
+// Heap-allocatable TryCatch
+namespace {
+struct _TryCatch {
+  explicit _TryCatch(v8::Isolate* isolate) : real_try_catch(isolate) {}
 
-MaybeLocal<Value> MakeCallback(Environment* env,
-                               Local<Value> recv,
-                               const Local<Function> callback,
-                               int argc,
-                               Local<Value> argv[],
-                               async_context asyncContext) {
-  // If you hit this assertion, you forgot to enter the v8::Context first.
-  CHECK_EQ(env->context(), env->isolate()->GetCurrentContext());
+  v8::TryCatch real_try_catch;
+};
+}
 
-  Local<Object> object;
+class InternalCallbackScope {
+ public:
+  InternalCallbackScope(Environment* env,
+                        Local<Object> object,
+                        const async_context& asyncContext);
+  ~InternalCallbackScope();
+  void Close();
 
-  Environment::AsyncCallbackScope callback_scope(env);
-  bool disposed_domain = false;
+  inline bool Failed() const { return failed_; }
+  inline void MarkAsFailed() { failed_ = true; }
+  inline v8::Local<v8::Value> GetFailureReturnValue() const {
+    return failure_return_value_;
```

Comment:
every case I can follow has a return value of `Undefined()`. is this extra logic necessary?

Reply:
Not anymore :+1: